code/datums/components/bayonet_charge.dm

```diff
@@ -0,0 +1,183 @@
+	. = ..()
+
+/datum/action/bayonetcharge/remove_action(mob/M)
+
+
+	parent.charge_off()
+	. = ..()
+
+/datum/action/bayonetcharge/action_activate()
+	if(parent.amcharging)
+		return
+	if(parent.chargewhenmove == TRUE)
+		parent.charge_off()
+		return
+
+	parent.charge_on()
+
+/datum/component/bayonetcharge/proc/charge_on()
+	if(incooldown)
+		to_chat(charger, "<span class='warning'>Charge is still on cooldown!</span>")
+		return
+	chargewhenmove = TRUE
+	child.add_selected_frame()
+	 //just in case you didnt activate the action, charger would be null if you do.
+	to_chat(charger, "<span class='warning'>You stand ready to charge</span>")
+	RegisterSignal(charger, COMSIG_MOVABLE_MOVED, .proc/update_charging)
+	RegisterSignal(charger, COMSIG_ATOM_DIR_CHANGE, .proc/on_dir_change)
+
+/datum/component/bayonetcharge/proc/charge_off()
+	chargewhenmove = FALSE
+
+	to_chat(charger, "<span class='warning'>You no longer stand ready to charge</span>")
+	if(charger) //locks up if you dont check if charger is existant
+		UnregisterSignal(charger, COMSIG_MOVABLE_MOVED)
+		UnregisterSignal(charger, COMSIG_ATOM_DIR_CHANGE)
+		UnregisterSignal(charger, list(COMSIG_MOVABLE_PREBUMP_TURF, COMSIG_MOVABLE_PREBUMP_MOVABLE, COMSIG_MOVABLE_PREBUMP_EXIT_MOVABLE))
+	child.remove_selected_frame()
+	stop_charge()
+
+
+/datum/component/bayonetcharge/proc/on_dir_change(datum/source, old_dir, new_dir)
+	SIGNAL_HANDLER
+	if(old_dir == new_dir)
+		return
+	stop_charge()
+//called when the marine moves
+/datum/component/bayonetcharge/proc/update_charging(datum/source, atom/oldloc, direction, Forced)
+	SIGNAL_HANDLER_DOES_SLEEP
+	if(movementcontinuationlimit && world.time > movementcontinuationlimit) //if you have stopped moving, it resetss the steps taken coutner
+		if(amcharging)
+			amcharging = FALSE
+			incooldown = TRUE
+			addtimer(CALLBACK(src, .proc/cooldownend), cooldowntime)
+		stop_charge()
+		movementcontinuationlimit = null
+		return
+	stepstaken++
+	movementcontinuationlimit = world.time + 10
+
+	handle_charge()
+
+/datum/component/bayonetcharge/proc/handle_charge()
+
+	//actual charging is under here
+	var/chargespeed = clamp(stepstaken * speedperstep, 0, maxspeed)
+	if(charger.incapacitated())
+		return
+
+	if(charger.pulling)
+		return
+	if(charger.getStaminaLoss() >= 0)
+		charge_off()
+		return
+	if(stepstaken < minactivationrundist)
+		return
+
+	//stuff in this if only happens once the charger starts charging, and not anytime else.
+	if(!amcharging)
+		charger.emote("warcry")
+		RegisterSignal(charger, list(COMSIG_MOVABLE_PREBUMP_TURF, COMSIG_MOVABLE_PREBUMP_MOVABLE, COMSIG_MOVABLE_PREBUMP_EXIT_MOVABLE), .proc/on_contact_with_enemy, TRUE)
+	amcharging = TRUE
+	charger.add_movespeed_modifier(MOVESPEED_ID_MARINE_CHARGE, TRUE, 100, NONE, TRUE, -chargespeed)
+	charger.toggle_move_intent(MOVE_INTENT_WALK) //if this aint here, the sprint compounds with the speed increase of the modifier, making you faster than a fucking charger
```

Comment:
ok, but can i cheat my way in an reactivate it during the charge?

Reply:
its called per step. this way the movespeed_modifier can keep up and ramp the speed up and so that you cant cheat by renabling it.

Reply:
ok